Improve handling of null and undefined Serverless variables and environment variables

skewart opened the pull request:
## What did you implement:

Closes #3491 

- Throws exception instead of logging a warning if a Serverless variable resolves to `undefined`.
- Allows `null` variables.
- Environment variables specified in `serverless.yml` that resolve to `null` are deleted from the environment objects and never passed into functions.  (Currently they are passed in to functions as the string `"null"`.)

## How did you implement it:

Fairly trivial, but still might as well explain...

I changed the `warnIfNotFound` function in `lib/classes/Variables.js` to `throwIfNotFound` and changed the logic slightly to throw an exception if the value being checked was `undefined` (but not if it's `null`).

I also added a `deleteNullEnvironmentVariables` function to `lib/classes/Variables.js`, which deletes keys pointing to `null` values from the provider and function environments.  If these keys-value pairs are not deleted the environments will contain env vars with the key name and the string `"null"` as the value.

## How can we verify it:

Use the following service...

serverless.yml:

```yaml
service: foo-service

provider:
  name: aws
  runtime: nodejs6.10
  environment:
    BAR_PATH: null
custom:
  defaults:
    foo_path: null

functions:
  hello:
    handler: handler.hello
    environment:
     FOO_PATH: ${opt:foo-path, env:FOO_PATH, self:custom.defaults.foo_path}
     BAZ_PATH: null

```

handler.js:

```javascript
'use strict';

const DEFAULT_FOO_PATH = 'data/foo/yo';

module.exports.hello = (event, context, callback) => {

  console.log("FOO_PATH really is undefined: " + (process.env.FOO_PATH === undefined));
  console.log("BAR_PATH really is undefined: " + (process.env.BAR_PATH === undefined));
  console.log("BAZ_PATH really is undefined: " + (process.env.BAZ_PATH === undefined));

};
```

1. Run `sls invoke local -f hello`
2. Observe console logs indicating all three environment variables mentioned are `undefined`.
3. Delete lines 8-10 (the `custom.defaults.foo_path` section) from the `serverless.yml` file.
4. Run `sls invoke local -f hello` again.
5. Observe an error was thrown and the function was never invoked.
6. With the `custom.defaults.foo_path` section still deleted, run `sls invoke local -f hello --foo-path foo/bar`
7. Observe no error is thrown this time, and the first console log says `FOO_PATH really is undefined: false` (you set that env var from the --foo-path CLI argument).

## Todos:

- [X] Write tests
- [x] Write documentation - n/a ?
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** YES (Kind of?  This could break code that was written to depend on the existing behavior - e.g. checking if an environment variable the string `"undefined"` instead of checking against _being_ `null`).

skewart commented:
@pmuens @HyperBrain this implements the changes we discussed in #3491.   Should I add a note to the docs too?   Maybe somewhere in the Variables guide?

skewart commented:
@pmuens I added a bit about undefined and null variables to the docs for all the providers.  Not sure if that sentence about environment variables is totally necessary - it kind of seems pretty intuitive - but just thought it might be worth explicitly stating.      I'm happy to add/remove/change any of the text.

HyperBrain commented on docs/providers/aws/guide/variables.md:
```diff
@@ -281,6 +281,9 @@ custom:
 In this example, we're overwriting the default regex for our variable syntax. So whenever you define variables, you now need to use `${{}}` instead of `${}` (double curly brackets).
 
+## Undefined and null Variables
+Variables that resolve to `undefined` will cause an exception to be thrown when deploying or invoking a function locally. Variables that resolve to `null` will not.  Environment variables with `null` values will not be set for the service.
```
Maybe `Environment variables explicitly set to *null* will be omitted from the deployed functions.` explains a bit better what happens exactly.

HyperBrain replied:
Maybe it has to be added with slight modifications for the other providers (I do not have enough experience with them to make a statement there ;-) )

HyperBrain commented on lib/classes/Variables.test.js:
```diff
@@ -76,6 +74,31 @@ describe('Variables', () => {
+
+      return serverless.variables.populateService().then(() => {
+        expect(serverless.service.provider.environment.SOME_VAR).to.equal('some-value');
+        expect(serverless.service.provider.environment.OTHER_VAR).to.equal(undefined);
```
I think `.to.be.undefined` will work here too.

HyperBrain commented on lib/classes/Variables.test.js:
```diff
@@ -242,6 +265,30 @@ describe('Variables', () => {
+      sinon.stub(serverless.variables, 'getValueFromSource').resolves(undefined);
+
+      return expect(serverless.variables.populateProperty(property, true))
+        .to.eventually.be.rejected;
```
You could even test for the correct rejection here: `.be.rejectedWith(...)`

HyperBrain commented on lib/classes/Variables.test.js:
```diff
@@ -875,67 +922,72 @@ describe('Variables', () => {
-      expect(logWarningSpy.calledOnce).to.be.equal(true);
-      expect(logWarningSpy.args[0][0]).to.contain('environment variable');
+      expect(() => serverless.variables.throwIfNotFound('env:service', undefined))
+        .to.throw(Error);
```
The block can be just be
```
expect(() => serverless.variables.throwIfNotFound('env:service', undefined))
       .to.throw(/environment variable/);
```

Same in the tests below. This eliminates the need of any try/catch blocks.

skewart commented:
@HyperBrain thanks for the feedback! 

It looks like Serverless isn't actually configuring any kind of environment variables for Azure, Google Cloud Functions, or Open Whisk, so I removed the sentence about null env vars from the docs for those providers.   Correct me if I'm wrong about that - I searched the code for "environment" and only saw `.envrionmennt` objects from `serverless.yml` being set for AWS - though it's kind of late here and maybe I missed something.

skewart commented on lib/classes/Variables.test.js:
```diff
@@ -76,6 +74,33 @@ describe('Variables', () => {
+      };
+
+      return serverless.variables.populateService().then(() => {
+        /* eslint-disable no-unused-expressions */
```
Using `.to.be.undefined` instead of `.to.equal(undefined)` caused eslint to raise a `no-unused-expressions` error.   This fixes the failure.  Though, I'm not sure if it's the best way to handle it.

It might not be bad to disable that particular check on all `.test.js` files, since it's not unlikely to pop up elsewhere in tests as long as you're using Chai.

HyperBrain replied:
@pmuens I like the idea of disabling the test for the unit tests. When using chai, there are a lot of cases where you call a property to test for a specific value.
Alternatively we could use 
```
.then(() => BbPromise.all([
  expect(...).to.be.undefined,
  expect(...).to.be.not.empty,
  ...
]));
```
That way the warning can be enabled and it will additionally work with expectations that return promises or values.

skewart replied:
Interesting.  Yeah, to your point, I'd guess there are likely to be more false positives for `no-unused-expressions` than actual positives in test files.   And, just as a general principal, it's better to configure your linter to work with how you want your code to be, as opposed to changing your code to fit the default constraints imposed by your linter.  

All that said, I might be underestimating the value of `no-unused-expressions` in test files, and `.all()`-ing the `expects` doesn't seem like a terrible workaround.

pmuens replied:
Yes, I agree 👍. disabling that for all tests in the linter config sounds like a good idea (also ran into this a bunch of times).

HyperBrain commented:
@skewart Maybe you're right. It seems that only AWS currently uses the environment variables in such a way. @pmuens Could you clarify?

pmuens commented on lib/classes/Variables.js:
```diff
@@ -344,11 +344,24 @@ class Variables {
   }
 
-  warnIfNotFound(variableString, valueToPopulate) {
+  deleteNullEnvironmentVariables() {
```
Still not 100% sure if this should be kept here since this is not Serverless Variable specific (and this file only contains the logic for the Serverless Variables system).

I would see this somewhere in the `package` or `deploy` plugin (not the AWS one though).

skewart replied:
Yeah, tbh, I did have second thoughts about putting this here when I was implementing it - it does seem like a different concern.

I'll take a stab at moving it to the `package` or `deploy` plugins.

pmuens replied:
Thanks for getting back @skewart 👍 

That sounds like a good idea. @HyperBrain what do you thing about it? In which plugin should be put it?

HyperBrain replied:
@pmuens Currently the variable evaluation is done in the package phase, so it should be long there for now, imo.

pmuens replied:
@HyperBrain thanks for getting back.

Yes, I agree. Putting it into the packaging phase sounds like the best idea! 👍

skewart commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-unused-expressions */
```
So, eslint only started supporting glob-based rule overrides (i.e. disable a specific rule in all files that match a glob) very recently (v4.x).  `eslint-airbnb-config` hasn't been updated to work with eslint v4.  There's [an issue on airbnb/javascript to add eslint v4 support](https://github.com/airbnb/javascript/issues/1447) and it seems people are working on it.  But it's not done yet.

I think it would be great to add the override to `.eslintrc.js` once `eslint-airbnb-config` supports eslint v4.  But in the mean time these file-wide disables will work.

HyperBrain replied:
Good investigation. Fully agree with that approach and switch to eslint v4 and a fixed template as soon as it is available.

skewart commented on lib/plugins/package/package.js:
```diff
@@ -66,6 +67,9 @@ class Package {
         .then(this.packageService),
 
+      'before:package:compileFunctions': () => BbPromise.bind(this)
+        .then(this.deleteNullEnvVars),
```
This seemed like the best place to hook in to things and delete the null env vars.  That said, I could see the logic in doing it at the very beginning.

skewart commented:
@pmuens @HyperBrain it's been a crazy busy week, but I finally made the changes we talked about.  Null env vars are now deleted in the package plugin instead of `Variables.js`.   Also, it turns out it's not quite possible to disable an eslint rule for all test files just yet (see my comment above).

HyperBrain commented on lib/plugins/package/package.js:
```diff
@@ -75,6 +79,19 @@ class Package {
     };
   }
+
+  deleteNullEnvVars() {
```
This could be put into a separate file in the lib folder, so that the index file only does the initialization of the package plugin and the action implementations are outsourced (like it is done for the other actions - packageService and zipService).

skewart commented:
@HyperBrain I moved `deleteNullEnvVars` to its own file in the `lib` dir, like `packageService` and `zipService`.

I also made one other super tiny change:  I moved the `eslint-disable` comments that I added to the test down a few lines to be after the `'use strict';` statements, since I noticed things were in that order in a few other places in the code base, and it's always nice to be consistent.

skewart commented:
@HyperBrain @pmuens just wanted to check in and see if you're looking for any other changes to this PR (or is the timing just not right to merge it)?

pmuens commented:
> @HyperBrain @pmuens just wanted to check in and see if you're looking for any other changes to this PR (or is the timing just not right to merge it)?

Hey @skewart thanks for getting back and pinging 💯 and sorry for the slow response on our end 😞.

We were pretty busy with products we just released at our Emit conference this morning.

This PR is definitely in our "In review" pipeline and will get some attention soon! We're currently following-up on the PRs and reviewing them! 👍

skewart commented:
I totally understand!  (Just wanted to make sure it wasn't blocked on my end - certainly sounds like it's not.)

Emit conference looked great btw, - at lest from what I could tell on Twitter.   And the Event Gateway stuff looks pretty cool.  Stoked to give it a spin.

pmuens commented:
> I totally understand! (Just wanted to make sure I wasn't blocking things on my end - certainly sounds like that's not the case.)

Thanks for getting back @skewart 👍. Yes, you're definitely not blocking anything here 😊.

> Emit conference looked great btw, - at lest from what I could tell on Twitter. And the Event Gateway stuff looks pretty cool. Stoked to give it a spin.

Thanks for the kind words and the feedback 👍! Definitely give it a spin and let us know if you have any questions / need any help with that!

HyperBrain commented:
Just a brainwave: What if we invent something like linters or compilers do? A switch like `--warn-as-error` (global switch) that can be evaluated by anyone (core + plugins) to switch warnings to exceptions internally?
That way it would be non-breaking, as the default would be to stay with the warnings. On CI/CD systems or for local testing (or just because of paranoia) one would enable the switch to be sure to catch issues throughout the whole package/deploy pipeline.

skewart commented:
Yeah, this does seem like the kind of change that would ideally be released with a major version change.   If you have `if (process.env.SOME_VAR === "undefined") { ...` somewhere in your code and then you update to a version with this change your code will break.    That said, I have no idea how many people are actually doing that, and this does seem like a good change to make to the framework sooner rather than later, and I have no idea when you're planning to release 2.0.  Overall though I do think it's better to be a bit conservative in situations like this.  So I'd vote to hold off until 2.0.

Adding a `--warn-as-error` flag in the mean time , like @HyperBrain suggested, seems like a really good idea.   The only drawback I could see is if there are (or soon will be) a bunch of warnings that people would be unlikely to want to treat as errors.   A quick search of the code shows that this undefined variable warning is the only warning there is (at least it's the only one created by calling `Error.logWarning()`).     If that is indeed the case then I can't see a downside to adding `--warn-as-error`.

horike37 commented:
Closing this issue for now since it took for a long time without any activity.
Feel free to re-open if you want to restart working on this.

skewart commented:
Hi @carlbergman, I'm not sure what Serverless folks plans are, but I don't think I'll have time to continue working on this any time soon.
